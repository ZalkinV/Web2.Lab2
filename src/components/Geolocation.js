import React from "react";
import { connect } from "react-redux";

import { fetchWeatherByCoords } from "../middlewares";
import { setGeolocationWeather, fetchWeatherPending, fetchWeatherSuccess, fetchWeatherError } from "../actions";


class Geolocation extends React.Component {
  render() {
    return (
      <div>
        <button
          onClick={() => this.handleClick()}
        >Get geolocation</button>
      </div>
    );
  }

  handleClick() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition((position) => this.props.fetchWeatherByCoords(position.coords));
    } else {
      alert("Your browser does not support geolocation!");
    }
  }
}


function mapStateToProps(state) {
  return {
    geolocation: state.geolocation
  };
}

function mapDispatchToProps(dispatch) {
  return {
    fetchWeatherByCoords: (coords) => {
      dispatch(fetchWeatherPending());
      
      let promise = fetchWeatherByCoords(coords);
      promise
        .then(response => {
          response.json()
            .then(json => {
              console.log(response, json);
              if (!response.ok) {
                dispatch(fetchWeatherError(json.message));
              } else {
                dispatch(fetchWeatherSuccess());
                dispatch(setGeolocationWeather(json))
              }
            });
        },
        error => dispatch(fetchWeatherError(error)))
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Geolocation);
