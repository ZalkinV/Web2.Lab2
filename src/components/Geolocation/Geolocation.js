import React from "react";
import { connect } from "react-redux";

import { setGeolocation, fetchWeatherByCoords } from "../../actions/geoActions";
import Weather from "../Weather/Weather";

import "./Geolocation.css";


class Geolocation extends React.Component {
  render() {
    return (
      <div className="geolocation">
        <h1 className="header">Refresh geolocation</h1>
        <button className="button"
          onClick={() => this.handleClick()}
        >Get geolocation</button>
        <Weather
          onFetch={() => this.props.fetchWeatherByCoords(this.props.coords)}
          forecast={this.props.forecast}/>
      </div>
    );
  }

  handleClick() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => {
        const coords = {
          lat: position.coords.latitude,
          lon: position.coords.longitude
        };
        this.props.setGeolocation(coords);

        this.props.fetchWeatherByCoords(coords);
      });
    } else {
      alert("Your browser does not support geolocation!");
    }
  }
}


function mapStateToProps(state) {
  return {
    coords: state.geo.coords,
    forecast: state.geo.forecast
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setGeolocation: (coords) => {
      dispatch(setGeolocation(coords));
    },

    fetchWeatherByCoords: (coords) => {
      dispatch(fetchWeatherByCoords(coords));
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Geolocation);
