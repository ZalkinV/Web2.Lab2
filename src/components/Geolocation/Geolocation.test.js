import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Geolocation from "./Geolocation";


describe("<Geolocation /> render", () => {

  const DEFAULT_FORECAST = {
    cityName: "New York",
    temperature: "-22",
    icon: "icon_code",
    windSpeed: 15,
    description: "good to stay at home",
    pressure: 234,
    humidity: 1,
    coords: {
      lat: 74,
      lon: 89
    }
  };

  const DEFAULT_COORDS = {
    lat: 12,
    lon: 34
  };

  const storeCreator = configureStore();


  test("should has only coords", () => {
    const store = storeCreator({
      geo: {
        coords: DEFAULT_COORDS
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Geolocation
            onFetch={() => {}}
          />
        </Provider>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has coords and forecast", () => {
    const store = storeCreator({
      geo: {
        coords: DEFAULT_COORDS,
        forecast: DEFAULT_FORECAST
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Geolocation
            onFetch={() => {}}
          />
        </Provider>)
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

});
