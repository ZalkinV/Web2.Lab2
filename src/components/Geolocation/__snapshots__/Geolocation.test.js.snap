// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Geolocation /> render should has only coords 1`] = `
<div
  className="geolocation"
>
  <h1
    className="header"
  >
    Refresh geolocation
  </h1>
  <button
    className="button"
    onClick={[Function]}
  >
    Get geolocation
  </button>
</div>
`;
